### ISSUE TITLE
TrackedReal error

### ISSUE THREAD

**Issue by ghiro90** [FluxML/Flux.jl#800]
Hi all, I am  trying to implement this simple model in Flux

```using Flux
using Flux.Tracker
using Flux.Tracker: update!
using LinearAlgebra

x = rand(200,50)
y = rand(200,1)
w = rand(50,1)

λ = 20

predict(x) = x*w

penalty() = norm(w,2)

function loss(x, y, λ)
  ŷ = predict(x)
  sum((y .- ŷ).^2) + λ*penalty()
end

w = param(w)

opt = Descent(0.1)
data = [(x,y)]

Flux.train!(loss,w,data,opt)
```

however, I got following error
```
type TrackedReal has no field grad
getproperty(::Any, ::Symbol) at sysimg.jl:18
_update_params! at train.jl:11 [inlined]
macro expansion at train.jl:76 [inlined]
macro expansion at progress.jl:124 [inlined]
#train!#5(::getfield(Flux.Optimise, Symbol("##7#11")), ::Function, ::Function, ::TrackedArray{…,Array{Float64,2}}, ::Array{Tuple{Array{Float64,2},Array{Float64,2}},1}, ::Descent) at train.jl:72
train!(::Function, ::TrackedArray{…,Array{Float64,2}}, ::Array{Tuple{Array{Float64,2},Array{Float64,2}},1}, ::Descent) at train.jl:70
top-level scope at none:0
include_string(::Module, ::String, ::String) at loading.jl:1008
include_string(::Module, ::String, ::String, ::Int64) at eval.jl:30
(::getfield(Atom, Symbol("##114#119")){String,Int64,String})() at eval.jl:94
withpath(::getfield(Atom, Symbol("##114#119")){String,Int64,String}, ::Nothing) at utils.jl:30
withpath at eval.jl:46 [inlined]
#113 at eval.jl:93 [inlined]
with_logstate(::getfield(Atom, Symbol("##113#118")){String,Int64,String}, ::Base.CoreLogging.LogState) at logging.jl:395
with_logger at logging.jl:491 [inlined]
#112 at eval.jl:92 [inlined]
hideprompt(::getfield(Atom, Symbol("##112#117")){String,Int64,String}) at repl.jl:85
macro expansion at eval.jl:91 [inlined]
(::getfield(Atom, Symbol("##111#116")))(::Dict{String,Any}) at eval.jl:86
handlemsg(::Dict{String,Any}, ::Dict{String,Any}) at comm.jl:164
(::getfield(Atom, Symbol("##19#21")){Array{Any,1}})() at task.jl:259
```
Thanks in advance for any advice.

**Comment by thebhatman** [comment]
It's always better to define your own training loop. Flux.train!() is very basic and pretty specific. I got it working with this piece of code, by calculating gradients and updating parameters of the model : 
```
using Flux.Tracker
using Flux.Tracker: update!
using LinearAlgebra

x = rand(200,50)
y = rand(200,1)
w = rand(50,1)

λ = 20

predict(x) = x*w

penalty() = norm(w,2)

function loss(x, y, λ )
  ŷ = predict(x)
  param(sum((y .- ŷ).^2) + λ*penalty())
end

w1 = params(w)

opt = Descent(0.1)
data = [(x,y)]
for d in data
  grads = Tracker.gradient(() -> loss(d..., λ), w1)
  update!(opt, w1, grads)
end
```
Do note that `param` creates only a tracked object, while `params` is used for parameters of the model.

**Comment by ghiro90** [comment]
Thank you, I adjusted the code as follows
``` 
function lm_srhink(x,y,λ,η,l)


### Compute the Lasso
# estimates are obtained by gradient descent

# x: input matrix
# y: response vector
# λ: shrinkage parameter
# η: learning rate
# l: norm to be used


# check for the input
w = param(transpose(mean(x,dims = 1)))

# define the linear model
ln(x) = x*w

#define the penalty function
penalty() = norm(w,l)

# loss function to minimize
    function ℓ(x, y)
    ŷ = ln(x)
        sum((y .- ŷ).^2) + λ*penalty()
    end

#set parameters for gradient descnet
w1 = params(w)

# Define learning rate for the
opt = Descent(η)

ε = 10^9

    # da cambiare con il while
    while ε > 0.05
        Δ = Tracker.gradient(() -> ℓ(x,y), w1)
        update!(opt,w1,Δ)
        global ε += abs(ε - ℓ(x,y))
        end
       
return w, ℓ(x,y)
    end
```

however now it returns that it can not evaluate the gradient because the loss function is infinite.
I am using L1 norm penality, lambda = 30 and eta = 0.1.

**Comment by thebhatman** [comment]
Your ε  variable is continuously increasing.... It's an infinite loop. I just changed `global ε -= abs(ε - ℓ(x,y))` the plus to minus and it works.
